core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -203,20 +204,22 @@ protected void add(DataFile file) {
    * Add all files in a manifest to the new snapshot.
    */
   protected void add(ManifestFile manifest) {
-    // the manifest must be rewritten with this update's snapshot ID
-    try (ManifestReader reader = ManifestReader.read(
-        ops.io().newInputFile(manifest.path()), ops.current().specsById())) {
-      ManifestFile manifestFile = ManifestWriter.copyAppendManifest(
-          reader, manifestPath(manifestCount.getAndIncrement()), snapshotId(), appendedManifestsSummary);
-      appendManifests.add(manifestFile);
-      // keep reference of the first appended manifest, so that we can avoid merging first bin(s)
-      // which has the first appended manifest and have not crossed the limit of minManifestsCountToMerge
-      if (firstAppendedManifest == null) {
-        firstAppendedManifest = manifestFile;
+    // TODO: avoid reading manifests to simply get stats
+    try (ManifestReader reader = ManifestReader.read(manifest, ops.io(), ops.current().specsById())) {
+      for (ManifestEntry addedEntry : reader.addedFiles()) {
+        appendedManifestsSummary.addedFile(reader.spec(), addedEntry.file());
       }
     } catch (IOException e) {
       throw new RuntimeIOException(e, "Failed to close manifest: %s", manifest);
     }
+
+    appendManifests.add(manifest);
```

Comment:
Why not copy the `ManifestFile` with the snapshot ID here instead of adding a new loop in `apply`?

Reply:
The intent was to reuse the same mechanism and logic to set sequence numbers and any other metadata that will be known only in `apply`. That way, we don't have to copy `ManifestFile` twice.

Reply:
Can you add a comment in `apply` that explains this? Maybe just a `TODO: add sequence numbers here`.

Reply:
Done